k9mail/src/test/java/security/planck/ui/provisioning/ProvisioningViewModelTest.kt

```diff
@@ -1,98 +1,57 @@
+import org.junit.Rule
 import org.junit.Test
-import security.planck.provisioning.InitializationFailedException
+import security.planck.common.LiveDataTest
 import security.planck.provisioning.ProvisionState
-import security.planck.provisioning.ProvisioningFailedException
 import security.planck.provisioning.ProvisioningManager
 
-class ProvisioningViewModelTest {
-    private val provisioningManager: ProvisioningManager = mockk(relaxed = true)
-    private val view: ProvisioningView = mockk(relaxed = true)
-    private val viewModel = ProvisioningViewModel(provisioningManager)
+@OptIn(ExperimentalCoroutinesApi::class)
+class ProvisioningViewModelTest : LiveDataTest<ProvisionState>() {
+    @get:Rule
+    var coroutinesTestRule = CoroutineTestRule()
 
-    @Test
-    fun `view handles WaitingForProvisioning state`() {
-        viewModel.attach(view)
-
-
-        viewModel.provisionStateChanged(ProvisionState.WaitingForProvisioning)
-
-
-        verify { view.waitingForProvisioning() }
-    }
-
-    @Test
-    fun `view handles InProvisioning state`() {
-        viewModel.attach(view)
-
-
-        viewModel.provisionStateChanged(ProvisionState.InProvisioning)
-
-
-        verify { view.provisioningProgress() }
-    }
-
-    @Test
-    fun `view handles Initializing state after successful provisioning`() {
-        viewModel.attach(view)
-
-
-        viewModel.provisionStateChanged(ProvisionState.Initializing(true))
-
-
-        verify { view.initializingAfterSuccessfulProvision() }
-    }
-
-    @Test
-    fun `view handles Initializing state`() {
-        viewModel.attach(view)
-
-
-        viewModel.provisionStateChanged(ProvisionState.Initializing(false))
-
-
-        verify { view.initializing() }
+    private val testFlow =
+        MutableStateFlow<ProvisionState>(ProvisionState.WaitingToInitialize(false))
+    private val provisioningManager: ProvisioningManager = mockk(relaxed = true) {
+        every { state }.returns(testFlow)
     }
-
-    @Test
-    fun `view handles Initialized state`() {
-        viewModel.attach(view)
-
-
-        viewModel.provisionStateChanged(ProvisionState.Initialized)
-
-
-        verify { view.initialized() }
+    private lateinit var viewModel: ProvisioningViewModel
+    override val testLivedata: LiveData<ProvisionState>
+        get() = viewModel.state
+
+    override fun initialize() {
+        //Dispatchers.setMain(coroutinesTestRule.testDispatcher)
+        viewModel = ProvisioningViewModel(
+            provisioningManager,
+        )
     }
 
     @Test
-    fun `view handles provisioning Error state`() {
-        viewModel.attach(view)
-
-
-        viewModel.provisionStateChanged(
-            ProvisionState.Error(ProvisioningFailedException("test error", RuntimeException()))
+    fun `view emits state from ProvisioningManager`() = runTest {
+        assertObservedValues(
+            ProvisionState.WaitingToInitialize(false),// initial state
+            ProvisionState.WaitingToInitialize(false),
```

Comment:
two of the same line? Can you clarify?

Reply:
First is the initial value of the LiveData in the ViewModel, second is the initial state coming from the ProvisioningManager Flow.
I gave an initial value to LiveData so it would never be null.